packages/ember-application/lib/system/application-instance.js

```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
     }
   },
 
-  router: computed(function() {
-    return this.lookup('router:main');
-  }).readOnly(),
+  _bootOptions: null,
+  _bootPromise: null,
+  _booted: false,
 
   /**
-    Instantiates and sets up the router, specifically overriding the default
-    location. This is useful for manually starting the app in FastBoot or
-    testing environments, where trying to modify the URL would be
-    inappropriate.
+    Initialize the `Ember.ApplicationInstance` and return a promise that resolves
+    with the instance itself when the boot process is complete.
+
+    The primary task here is to run any registered instance initializers.
 
+    See the documentation on `BootOptions` for the options it takes.
+
+    @private
+    @method boot
     @param options
+    @return {Promise<Ember.ApplicationInstance,Error>}
+  */
+  boot(options) {
+    if (this._bootPromise) { return this._bootPromise; }
+
+    let defer = new Ember.RSVP.defer();
+    let promise = this._bootPromise = defer.promise;
+
+    try {
```

Comment:
Why use a deferred here? Isn't this the re-implementing the same semantics as if you wrapped it in a `Promise`?